audits/trading/202404-threat-model-immutable-signed-zone-v2.md

```diff
@@ -2,14 +2,214 @@
+| `revokeRole(bytes32,address)`    | d547741f          | Role admin      |
+| `updateAPIEndpoint(string)`      | 297234d7          | `DEFAULT_ADMIN` |
+| `updateDocumentationURI(string)` | 0a904f08          | `DEFAULT_ADMIN` |
+
+Functions that *do not change* state:
+
+| Name                                                                                                                                                           | Function Selector |
+| -------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------- |
+| `DEFAULT_ADMIN_ROLE()`                                                                                                                                         | a217fddf          |
+| `getRoleAdmin(bytes32)`                                                                                                                                        | 248a9ca3          |
+| `getRoleMember(bytes32,uint256)`                                                                                                                               | 9010d07c          |
+| `getRoleMemberCount(bytes32)`                                                                                                                                  | ca15c873          |
+| `getSeaportMetadata()`                                                                                                                                         | 2e778efc          |
+| `hasRole(bytes32,address)`                                                                                                                                     | 91d14854          |
+| `sip7Information()`                                                                                                                                            | d600940e          |
+| `supportsInterface(bytes4)`                                                                                                                                    | 01ffc9a7          |
+| `validateOrder((bytes32,address,address,(uint8,address,uint256,uint256)[],(uint8,address,uint256,uint256,address)[],bytes,bytes32[],uint256,uint256,bytes32))` | 17b1f942          |
+
+### Admin Roles
+
+Accounts with administrative privileges could be used by attackers to facilitate attacks. This section analyses what each role can do.
+
+#### Accounts with `DEFAULT_ADMIN` role on ImmutableSignedZoneV2 contract
+
+This role is granted to the `owner` specified in the `constructor` of the contract. Accounts with the `DEFAULT_ADMIN` account can:
+
+* Grant can grant administrator roles to any account, including the `DEFAULT_ADMIN` role
+* Revoke `DEFAULT_ADMIN` role from any account
+* Renounce the `DEFAULT_ADMIN` role for itself, possibly leading to no administrators and loss of control of the contract
```

Comment:
Should the Renounce function be stopped from removing the last DEFAULT_ADMIN_ROLE? Assets have / are implementing this style of function in their code

Reply:
actioned

Reply:
Updates: `revokeRole` and `renounceRole` are now protected from removing the last `DEFAULT_ADMIN_ROLE`.